File: src/lib/code-index.ts
/**
 * Code index types and cache for .NET (granit-dotnet) and
 * TypeScript (granit-front) source code.
 *
 * Indexes are generated at CI time — by RoslynLens for .NET,
 * by ts-morph for TypeScript — and published as GitHub Release Assets.
 */

import type { KVCache } from './index-cache.js';
import { getIndex } from './index-cache.js';

// ─── .NET code index (code-index.json) ────────────────────────────────────────

export interface CodeMember {
  name: string;
  kind: string;       // "method" | "property" | "event" | "field"
  signature: string;
  returnType?: string;
}

export interface CodeSymbol {
  name: string;
  fqn: string;        // Fully-qualified name
  kind: string;       // "class" | "interface" | "enum" | "record" | "struct"
  project: string;
  file: string;
  line?: number;
  members: CodeMember[];
}

export interface ProjectNode {
  name: string;
  deps: string[];
  framework: string;
}

export interface CodeIndex {
  generatedAt: string;
  repo: string;
  projectGraph: ProjectNode[];
  symbols: CodeSymbol[];
}

// ─── TypeScript front index (front-index.json) ───────────────────────────────

export interface FrontExport {
  name: string;
  kind: string;       // "function" | "interface" | "type" | "class" | "enum"
  signature: string;
  members?: CodeMember[];
}

export interface FrontPackage {
  name: string;
  description: string;
  exports: FrontExport[];
}

export interface FrontIndex {
  generatedAt: string;
  repo: string;
  packages: FrontPackage[];
}

// ─── Cache helpers ────────────────────────────────────────────────────────────

export async function getCodeIndex(
  url: string,
  cache: KVCache,
): Promise<CodeIndex | null> {
  try {
    return await getIndex<CodeIndex>('code', url, cache);
  } catch {
    return null;
  }
}

export async function getFrontIndex(
  url: string,
  cache: KVCache,
): Promise<FrontIndex | null> {
  try {
    return await getIndex<FrontIndex>('front', url, cache);
  } catch {
    return null;
  }
}

File: src/Granit.Mcp/Models/CodeIndex.cs
using System.Text.Json.Serialization;

namespace Granit.Mcp.Models;

public sealed record CodeIndex(
    [property: JsonPropertyName("repo")] string Repo,
    [property: JsonPropertyName("projectGraph")] List<ProjectNode> ProjectGraph,
    [property: JsonPropertyName("symbols")] List<CodeSymbol> Symbols);

public sealed record ProjectNode(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("deps")] List<string> Deps,
    [property: JsonPropertyName("framework")] string Framework);

public sealed record CodeSymbol(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("fqn")] string Fqn,
    [property: JsonPropertyName("kind")] string Kind,
    [property: JsonPropertyName("project")] string Project,
    [property: JsonPropertyName("file")] string File,
    [property: JsonPropertyName("line")] int? Line,
    [property: JsonPropertyName("members")] List<CodeMember> Members);

public sealed record CodeMember(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("kind")] string Kind,
    [property: JsonPropertyName("signature")] string Signature,
    [property: JsonPropertyName("returnType")] string? ReturnType);
